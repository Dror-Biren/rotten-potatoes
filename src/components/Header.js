import React from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';

import { startLogin } from '../actions/auth';
import { startLogout } from '../actions/auth';
import { history } from '../routers/AppRouter';

export const Header = ({ isAuthenticated, startLogin, startLogout }) => {
   const logOutButton = (
      <button className="button" onClick={startLogout}>
         Log-out
      </button>
   );

   const logInButton = (
      <button className="button" onClick={startLogin}>
         Login with Google
      </button>
   );

   return (
      <header className="header">
         <img
            className="header__background"
            src="/images/potatoes.webp"
         />
         <div className="content-container">
            <div className="header__content">

               <Link className="header__title" to="/dashboard">
                  <h1>
                     Rotten Potatoes
                  </h1>
               </Link>

               { isAuthenticated ? logOutButton : logInButton }
            </div>
         </div>
      </header>
   );
}

const mapStateToProps = (state) => ({
   isAuthenticated: !!state.auth.uid
 });

const mapDispatchToProps = (dispatch) => ({
   startLogin: () => dispatch(startLogin()),
   startLogout: () => {
      dispatch(startLogout()).then();
      const returnToLoginPage = () => history.push('/');
      setTimeout(returnToLoginPage, 700);
   }
});

export default connect(mapStateToProps, mapDispatchToProps)(Header);
