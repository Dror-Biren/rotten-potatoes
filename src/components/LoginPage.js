import React from 'react';
import { connect } from 'react-redux';
import { history } from '../routers/AppRouter';
import { Redirect } from 'react-router-dom';

import { startLogin } from '../actions/auth';
import store from '../store/configureStore';


const goToDashboardPage = () => {
   history.push('/dashboard')
}

export class LoginPage extends React.Component {
   constructor(props) {
      super(props);

      this.state = {
         classNames: "",
         animationStarted: false,
         authenticated: false
      };
   }

   componentDidMount() {
      setTimeout(() => {
         const state = store.getState();
         console.log({ state });
         if (state.auth.uid)
            this.setState({authenticated: true});
      }, 200) 
   }

   shiftPageAnimation = () => {
      this.setState({
         classNames: "animation",
         animationStarted: true
      });
   };

   shiftPageAfterLogin = () => {
      this.props.startLogin(this.shiftPageAnimation);
   }

   render() {
      if (this.state.authenticated)
         return <Redirect to="/dashboard" />

      const title = (
         <h1 className="login-ui__title">
            Rotten
            &#10;
            Potatoes
         </h1>
      )

      const subTitle = (
         <p>
            We rate movies. Join us!
         </p>
      );

      const loginButton = (
         <button className="button" onClick={this.shiftPageAfterLogin}>
            Login with Google
         </button>
      );

      const textBetweenButtons = (
         <p>
            ------- OR -------
         </p>
      )

      const continueAsGuestButton = (
         <button className="button continueAsGuest" onClick={this.shiftPageAnimation}>
            Continue as guest
         </button>
      );


      return (
         <div className="login-ui__container">
            <div
               className={`login-ui ${this.state.classNames}`}
               onAnimationEnd={goToDashboardPage}>
               {title}

               {
                  !this.state.animationStarted &&
                  <div>
                     {subTitle}
                     <div className="loginButtons">
                        {loginButton}
                        {textBetweenButtons}
                        {continueAsGuestButton}
                     </div>
                     <button className="startAnimation" onClick={this.shiftPageAnimation}>
                        start animation
                     </button>
                  </div>
               }
            </div>
         </div >
      );
   }
}

const mapDispatchToProps = (dispatch) => ({
   startLogin: (callback) => dispatch(startLogin(callback))
});

export default connect(undefined, mapDispatchToProps)(LoginPage);
