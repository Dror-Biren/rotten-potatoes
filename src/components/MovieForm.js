import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';


import Poster from './Poster';
import SelectGenres3 from './SelectGenres3';
import { moviePramsMax } from '../appConsts';
//import MoviePopUp from './MoviePopUp';
//console.log(MoviePopUp)


export default class MovieForm extends React.Component {

   constructor(props) {
      super(props);

      this.state = {
         title: props.movie ? props.movie.title : '',
         description: props.movie ? props.movie.description : '',
         releaseDate: props.movie ? moment(props.movie.releaseDate) : moment(),

         trailerUrl: props.movie ? props.movie.trailerUrl : '',
         posterUrl: props.movie ? props.movie.posterUrl : null,
         poster: null,
         genres: props.movie ? props.movie.genres : [],

         calendarFocused: false,
         error: ''
      };
   }

   onTitleChange = (e) => {
      const title = e.target.value;
      this.setState(() => ({ title }));
   };

   onGenresChanged = (genres) => {
      this.setState(() => ({ genres }))
   }

   onDescriptionChange = (e) => {
      const description = e.target.value;
      this.setState(() => ({ description }));
   };

   onTrailerChange = (e) => {
      const trailerUrl = e.target.value;
      this.setState(() => ({ trailerUrl }));
   }

   onDateChange = (releaseDate) => {
      if (releaseDate) {
         this.setState(() => ({ releaseDate }));
      }
   };

   onFocusChange = ({ focused }) => {
      const focusedObj = { calendarFocused: focused }
      this.setState(() => focusedObj);
   };

   onSelectPoster = (event) => {
      const poster = event.target.files[0];

      if (poster) {
         if (poster.size > moviePramsMax.posterBytes)
            return alert("File is too big!");
         const posterUrl = URL.createObjectURL(poster);
         const posterObj = { poster, posterUrl };
         this.setState(() => posterObj);
      }
   }

   onRemovePoster = (e) => {
      e.preventDefault();

      this.setState(() => ({
         poster: null,
         posterUrl: null
      }));
   }

   onSubmit = (e) => {
      e.preventDefault();

      const { title, description, poster, posterUrl, trailerUrl, genres } = this.state;

      const setError = text => {
         const errorObj = { error: text };
         this.setState(() => errorObj);
      };

      if (!title)
         return setError('Please provide title.');

      if (genres.length === 0 || genres.length > moviePramsMax.genresAmount)
         return setError(`Please provide between 1-${moviePramsMax.genresAmount} genres.`);

      setError('');

      
      const releaseDate = this.state.releaseDate.valueOf();
      const movieData = { title, description, poster, releaseDate, trailerUrl, genres };
      if (!posterUrl)
         movieData.posterUrl = null;

      this.props.onSubmit(movieData);
   };


   render() {
      return (
         <form className="form" onSubmit={this.onSubmit}>

            {
               this.state.error &&
               <p className="form__error">
                  {this.state.error}
               </p>
            }

            <input
               type="text"
               placeholder="title"
               maxLength={moviePramsMax.titleLength}
               autoFocus
               className="text-input"
               value={this.state.title}
               onChange={this.onTitleChange}
            />

            <SelectGenres3
               onGenresChanged={this.onGenresChanged}
               initGenres={this.state.genres}
            />

            <Poster url={this.state.posterUrl} />

            <input
               type="file"
               id="poster"
               name="poster"
               accept="image/png, image/jpeg, image/jpg"
               onChange={this.onSelectPoster}
            ></input>

            <button
               onClick={this.onRemovePoster}
               disabled={!this.state.posterUrl}>
               Remove poster
            </button>


            <SingleDatePicker
               date={this.state.releaseDate}
               onDateChange={this.onDateChange}
               focused={this.state.calendarFocused}
               onFocusChange={this.onFocusChange}
               numberOfMonths={1}
               isOutsideRange={() => false}
            />

            <textarea
               placeholder="Add a description for your movie (optional)"
               className="textarea"
               value={this.state.description}
               onChange={this.onDescriptionChange}
            ></textarea>

            <input
               type="text"
               placeholder="paste here youtube trailer url"
               maxLength={moviePramsMax.descriptionLength}
               autoFocus
               className="text-input"
               value={this.state.trailerUrl}
               onChange={this.onTrailerChange}
            />

            <div>
               <button className="button">
                  Save Movie
               </button>
            </div>

         </form>
      )
   }
}


/*
     <label htmlFor="poster">
         Upload a poster image for the movie:
       </label>
*/
