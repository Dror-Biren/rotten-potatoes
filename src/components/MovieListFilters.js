import React from 'react';
import { connect } from 'react-redux';

import SelectGenres from './SelectGenres';
import YearsSlider from './YearsSlider';
import { movieGenres } from '../appConsts';
import { setTextFilter, setSortBy, setGenresfilter } from '../actions/filters';


export class MovieListFilters extends React.Component {

   onTextChange = (e) => {
      this.props.setTextFilter(e.target.value);
   };

   onSortChange = (e) => {
      this.props.setSortBy(e.target.value);
   };

   onGenresChanged = (genresSelected) => {
      //console.log(genres)
      this.props.setGenresfilter(genresSelected);
   }

   render() {


      return (
         <div className="page-margin itemsBackground filters">
            <div className="input-group">

               <div className="input-group__item small-sorters">
                  <input
                     type="text"
                     className="text-input"
                     placeholder="🔎  Search movie title"
                     value={this.props.filters.text}
                     onChange={this.onTextChange}
                  />

                  <select
                     className="select"
                     value={this.props.filters.sortBy}
                     onChange={this.onSortChange}
                  >
                     <option value="rating">Top rated ⭐</option>
                     <option value="viewes">Most viewed 👀</option>
                     <option value="date">Most new 🕥</option>
                  </select>
               </div>



               <SelectGenres
                  onGenresChanged={this.onGenresChanged}
                  initGenres={movieGenres}
               />


               <YearsSlider />
            </div>
         </div>
      );
   }
};

const mapStateToProps = (state) => ({
   filters: state.filters
});

const mapDispatchToProps = (dispatch) => ({
   setTextFilter: (text) => dispatch(setTextFilter(text)),
   setSortBy: (sorter) => dispatch(setSortBy(sorter)),
   setGenresfilter: (genres) => dispatch(setGenresfilter(genres))
});

export default connect(mapStateToProps, mapDispatchToProps)(MovieListFilters);


/*
<div className="input-group__item">
                  <select
                     className="select"
                     value={this.props.filters.sortBy}
                     onChange={this.onSortChange}
                  >
                     <option value="rating">Top rated ⭐</option>
                     <option value="viewes">Most viewed 👀</option>
                     <option value="date">Most new 🕥</option>
                  </select>
               </div>
*/