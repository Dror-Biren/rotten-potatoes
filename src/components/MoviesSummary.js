import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import selectMovies from '../selectors/movies';
import selectExpensesTotal from '../selectors/movies-total';

export const MoviesSummary = ({ moviesAmount, ratingsAmount }) => {
  return (
    <div className="page-header">
      <div className="content-container">
        <h1 className="page-header__title">
          Viewing <span>{ratingsAmount}</span> ratings, for <span>{moviesAmount}</span> movies.
        </h1>
        <h3>help us to rate them all!</h3>
        {
        <div className="page-header__actions">
          <Link className="button" to="/create">Add movie</Link>
        </div>
        }
      </div>
    </div>
  );
};

const mapStateToProps = (state) => {
  const visibleMovies = selectMovies(state.movies, state.filters);

  return {
    moviesAmount: visibleMovies.length,
    ratingsAmount: selectExpensesTotal(visibleMovies)
  };
};

export default connect(mapStateToProps)(MoviesSummary);
