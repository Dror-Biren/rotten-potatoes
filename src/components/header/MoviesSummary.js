import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';

import selectMovies from '../../selectors/movies';
import selectExpensesTotal from '../../selectors/movies-total';


export const MoviesSummary = ({ moviesAmount, ratingsAmount, isAdmin }) => {
   const adminAddMovieButton = (
      isAdmin &&
      <div className="button--add-movie">
         <Link className="add-movie-symbol" to="/create">
            +
         </Link>
      </div>
   );
   //âž•

   const spanBold = (text) => (
      <span className="span-bold">
         {text}
      </span>
   )

   const moviesSummaryText = (
      <h2 className="sub-header__title">
         Found {spanBold(moviesAmount)} movies,
         with a total of {spanBold(ratingsAmount)} ratings
      </h2>
   )

   const noMoviesText = (
      <h2 className="sub-header__title">
         No movie matches your search settings
      </h2>
   );

   return (
      <div className="sub-header">
         {moviesAmount > 0 ? moviesSummaryText : noMoviesText}     
         {adminAddMovieButton}
      </div>
   );
};

/*
 Viewing <span>{ratingsAmount}</span> ratings,
   for <span>{moviesAmount}</span> movies.

   <span id="span-moto">
                  help us to rate them all!
               </span>
*/

const mapStateToProps = (state) => {
   const visibleMovies = selectMovies(state.movies, state.filters);

   return {
      moviesAmount: visibleMovies.length,
      ratingsAmount: selectExpensesTotal(visibleMovies),
      isAdmin: state.user.isAdmin
   };
};

export default connect(mapStateToProps)(MoviesSummary);
