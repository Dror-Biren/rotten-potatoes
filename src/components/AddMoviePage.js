import React from 'react';
import { connect } from 'react-redux';
import MovieForm from './MovieForm';
import { startAddMovie } from '../actions/movies';

export class AddMoviePage extends React.Component {
   returnToMainPage = () => {
      this.props.history.push('/');
   };

   onSubmit = (movie) => {
      this.props.startAddMovie(movie);
      this.props.history.push('/');
   };

   render() {
      return (
         <div>
            <div className="sub-header">
               
                  <h2 className="sub-header__title">
                     Add a new movie to the site
                  </h2>
               
               <img
                  src="/images/goBack5.png"
                  className="button--go-back"
                  onClick={this.returnToMainPage}>
               </img>
            </div>
            <div className="page-margin">
               <MovieForm
                  onSubmit={this.onSubmit}
               />
            </div>
         </div>
      );
   }
}

const mapDispatchToProps = (dispatch) => ({
   startAddMovie: (movie) => dispatch(startAddMovie(movie))
});

export default connect(undefined, mapDispatchToProps)(AddMoviePage);
