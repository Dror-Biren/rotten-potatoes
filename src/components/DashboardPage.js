import React from 'react';

import MoviesList from './MoviesList';
import MovieListFilters from './MovieListFilters';
import MoviesSummary from './MoviesSummary';

const DashboardPage = () => (
  <div>
    <MoviesSummary />
    <MovieListFilters />
    <MoviesList />
  </div>
);

export default DashboardPage;
