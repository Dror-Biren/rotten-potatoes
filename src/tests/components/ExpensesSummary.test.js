import React from 'react';
import { shallow } from 'enzyme';
import { MoviesSummary } from '../../components/MoviesSummary';

test('should correctly render MoviesSummary with 1 movie', () => {
  const wrapper = shallow(<MoviesSummary movieCount={1} moviesTotal={235} />);
  expect(wrapper).toMatchSnapshot();
});

test('should correctly render MoviesSummary with multiple movies', () => {
  const wrapper = shallow(<MoviesSummary movieCount={23} moviesTotal={23512340987} />);
  expect(wrapper).toMatchSnapshot();
});
