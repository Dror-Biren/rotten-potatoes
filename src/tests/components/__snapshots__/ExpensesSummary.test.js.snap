// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should correctly render MoviesSummary with 1 movie 1`] = `
<div
  className="sub-header"
>
  <div
    className="page-margin"
  >
    <h1
      className="sub-header__title"
    >
      Viewing 
      <span>
        1
      </span>
       
      movie
       totalling 
      <span>
        $2.35
      </span>
    </h1>
    <div
      className="button--add-movie"
    >
      <Link
        className="button"
        replace={false}
        to="/create"
      >
        Add Movie
      </Link>
    </div>
  </div>
</div>
`;

exports[`should correctly render MoviesSummary with multiple movies 1`] = `
<div
  className="sub-header"
>
  <div
    className="page-margin"
  >
    <h1
      className="sub-header__title"
    >
      Viewing 
      <span>
        23
      </span>
       
      movies
       totalling 
      <span>
        $235,123,409.87
      </span>
    </h1>
    <div
      className="button--add-movie"
    >
      <Link
        className="button"
        replace={false}
        to="/create"
      >
        Add Movie
      </Link>
    </div>
  </div>
</div>
`;
